rust/gateway/src/eventloop.rs

```diff
@@ -0,0 +1,193 @@
+        }
+    }
+}
+
+impl Eventloop {
+    pub fn poll(&mut self, cx: &mut Context<'_>) -> Poll<Result<Infallible>> {
+        loop {
+            if let Poll::Ready(Some(ice_candidates)) = self.control_rx.poll_recv(cx) {
+                let _id = self.portal.send(
+                    "gateway",
+                    EgressMessages::BroadcastIceCandidates(ice_candidates),
+                );
+                continue;
+            }
+
+            match self.connection_request_tasks.poll_unpin(cx) {
+                Poll::Ready(((_, reference), Ok(Ok(gateway_rtc_session_description)))) => {
+                    let _id = self.portal.send(
+                        "gateway",
+                        EgressMessages::ConnectionReady(ConnectionReady {
+                            reference,
+                            gateway_rtc_session_description,
+                        }),
+                    );
+
+                    // TODO: If outbound request fails, cleanup connection.
+                    continue;
+                }
+                Poll::Ready(((client, _), Ok(Err(e)))) => {
+                    self.tunnel.cleanup_connection(client.into());
+                    let _ = self.tunnel.callbacks().on_error(&e);
+                    continue;
+                }
+                Poll::Ready(((client, reference), Err(e))) => {
+                    tracing::debug!(
+                        "Failed to establish connection {reference} from client {client:?}: {e}"
+                    );
+                    continue;
+                }
+                Poll::Pending => {}
+            }
+
+            match self.add_ice_candidate_tasks.poll_unpin(cx) {
+                Poll::Ready(Ok(Ok(()))) => {
+                    continue;
+                }
+                Poll::Ready(Ok(Err(e))) => {
+                    tracing::error!(err = ?e,"add_ice_candidate");
+                    let _ = self.tunnel.callbacks().on_error(&e);
+                    continue;
+                }
+                Poll::Ready(Err(e)) => {
+                    tracing::debug!("Failed to add ICE candidatee: {e}");
+                    continue;
+                }
+                Poll::Pending => {}
+            }
+
+            match self.portal.poll(cx)? {
+                Poll::Ready(phoenix_channel::Event::InboundMessage {
+                    msg: IngressMessages::Init(_),
+                    ..
+                }) => {
+                    todo!("received init after init");
+                }
+                Poll::Ready(phoenix_channel::Event::InboundMessage {
+                    msg: IngressMessages::RequestConnection(req),
+                    ..
+                }) => {
+                    let tunnel = Arc::clone(&self.tunnel);
+
+                    match self.connection_request_tasks.try_push(
+                        (req.client.id, req.reference.clone()),
+                        async move {
+                            tunnel
```

Comment:
We don't need to run this with a time out, since we're using ice trickle this resolves locally we shouldn't expect this to ever timeout.

Reply:
Yeah this is a tricky one. Two things:

1. I think having something bounded is always good.
2. `FuturesUnordered` can be "empty" and return you a `None`. Hence, you'll need a waker when it is `None` and trigger that when you push into the `FuturesUnordered`. We do that in `futures-bounded` automatically for you! Arguably, that is not an intrinsic concern of `futures-bounded` but happens to be convenient in this case. We can make our own utility in here that is unbounded but does the `Waker` dance for us but I am not sure it is worth it?

Reply:
Yeah, I think it's worth it, given that it solves 2. with the added benefit of being bounded it makes sense.

Reply:
> Yeah, I think it's worth it, given that it solves 2. with the added benefit of being bounded it makes sense.

Wait I am confused now. Do you want me to create our own abstraction or use `futures-bounded`?

Reply:
My bad, I just meant it makes sense using futures-bounded